ibis/backends/dask/execution/util.py

```diff
@@ -345,3 +348,116 @@ def safe_scalar_type(output_meta):
         output_meta = pd.Timestamp(1, tz=output_meta.tz, unit=output_meta.unit)
 
     return output_meta
+
+
+def add_partitioned_sorted_column(
+    df: Union[dd.DataFrame, dd.Series],
+    set_index: bool = False,
+    col_name: str = "_ibis_index",
+) -> Tuple[dd.DataFrame, List]:
+    """Add a column that is already partitioned and sorted
+    This columns acts as if we had a global index across the distributed data.
+    Important properties:
+    - Each row has a unique id (i.e. a value in this column)
+    - IDs within each partition are already sorted
+    - Any id in partition N_{t} is less than any id in partition N_{t+1}
+    We do this by designating a sufficiently large space of integers per
+    partition via a base and adding the existing index to that base. See
+    `helper` below.
+    Though the space per partition is bounded, real world usage should not
+    hit these bounds. We also do not explicity deal with overflow in the
+    bounds.
+
+    Parameters
+    ----------
+    df : dd.DataFrame
+        Dataframe to add a column to
+    set_index : bool, optional
+        Whether to also set that column as the index, by default False
+    col_name : str, optional
+        The column name, by default "_ibis_index"
+    Returns
+    -------
+    Tuple[dd.DataFrame, List]
+        Tuple of the new dataframe with column and divisions of that column
+    """
+    if isinstance(df, dd.Series):
+        df = df.to_frame()
+
+    if col_name in df.columns:
+        raise ValueError(f"Column {col_name} is already present in DataFrame")
+
+    def helper(
+        df: Union[dd.Series, dd.DataFrame],
+        partition_info: Dict[str, Any],
+        col_name: str,
+    ):
+        """Assigns a column with a unique id for each row"""
+        base = partition_info["number"] << 32
+        return df.assign(**{col_name: [base + idx for idx in df.index]})
+
+    original_meta = df._meta.dtypes.to_dict()
+    new_meta = {**original_meta, **{col_name: "int"}}
```

Comment:
int is int64? (maybe use "long" or "int64") In some systems "int" meant int32

Reply:
Pandas automatically treats this as `int64`. 

```python
>> pd.Series([1], dtype="int").dtype
dtype('int64')
```

Reply:
Right, pandas does that but for the reader of ibis, there are many things inside ibis that doesn't all treat "int" as int64, so wouldn't hurt to make that clear :)